pkg/binpack/efficiency.go

```diff
@@ -0,0 +1,156 @@
+// Copyright (c) 2019 Palantir Technologies. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package binpack
+
+import (
+	"math"
+
+	"github.com/palantir/k8s-spark-scheduler-lib/pkg/resources"
+)
+
+// AvgPackingEfficiency represents result packing efficiency per resource type for the entire cluster.
+// Computed as average packing efficiency over all node efficiencies.
+type AvgPackingEfficiency struct {
+	CPU    float64
+	Memory float64
+	GPU    float64
+	Max    float64
+}
+
+// LessThan compares two average packing efficiencies. For a single packing we take the highest of the
+// resources' efficiency. For example, when CPU is at 0.81 and Memory is at 0.54 the avg efficiency
+// is 0.81. One packing efficiency is deemed less efficient when its avg efficiency is lower than
+// the other's packing efficiency.
+func (p *AvgPackingEfficiency) LessThan(o AvgPackingEfficiency) bool {
+	return p.Max < o.Max
+}
+
+// WorstAvgPackingEfficiency returns a representation of a failed bin packing. Each individual resource
+// type is at worst possible (zero) packing efficiency.
+func WorstAvgPackingEfficiency() AvgPackingEfficiency {
+	return AvgPackingEfficiency{
+		CPU:    0.0,
+		Memory: 0.0,
+		GPU:    0.0,
+		Max:    0.0,
+	}
+}
+
+// PackingEfficiency represents result packing efficiency per resource type for one node. Computed
+// as the total resources used divided by total capacity.
+type PackingEfficiency struct {
+	NodeName string
+	CPU      float64
+	Memory   float64
+	GPU      float64
+}
+
+// Max returns the highest packing efficiency of CPU and Memory.
+func (p *PackingEfficiency) Max() float64 {
+	return math.Max(p.GPU, math.Max(p.CPU, p.Memory))
+}
+
+// ComputePackingEfficiencies calculates utilization for all provided nodes, given the new reservation.
+func ComputePackingEfficiencies(
+	nodeGroupSchedulingMetadata resources.NodeGroupSchedulingMetadata,
+	reservedResources resources.NodeGroupResources) map[string]*PackingEfficiency {
```

Comment:
I think you'll want to pass pointers to these?

Reply:
Do you mean `reservedResources`? The type `resources.NodeGroupResources` is defined as `type NodeGroupResources map[string]*Resources`.

`nodeGroupSchedulingMetadata` is a struct. The resource fields are pointers though.

Asking for a clarification.